woocommerce/woocommerce
# Product Collection: Enable Context-Aware Previews by Adding `usesReference` to `registerProductCollection`

### Changes proposed in this Pull Request:





Closes #47964 

In this PR, we are adding a new argument called `usesReference` to `registerProductCollection`. This argument will allow collections to specify the required context. When the required context isn't available on Editor side but will be available in Frontend, we will show a preview label.

Here is how `registerProductCollection` will look like:

```tsx
__experimentalRegisterProductCollection({
  ...,
  usesReference: ['product'],
)
```

`usesReference` can have following values:
- product
- archive
- cart
- order

`usesReference` will be used to implement multiple features in the future. Therefore, it's important to define what's part of this PR and what's not.

**What's part of this PR:**
- Adding `usesReference` argument to `registerProductCollection`.
- When reference isn't available on Editor side, but will be available on Frontend, we will show a preview label.
  - When `usesReference` is set to `product`, we will show a message "Actual products will vary depending on the product being viewed."
  - When `usesReference` is set to `archive`, we will show a message "Actual products will vary depending on the page being viewed."

**What's not part of this PR:**
- Linking a product with collection (#44877): When usesReference is set to product and product context isn't available, we will show a prompt to select a product. This way, product context will be manually selected. 
- Handling deleted product (#44878): When product context is selected manually and product is deleted, we will need to handle this case. Communicating with merchant that product is deleted and new product needs to be selected for collection to work properly.
- Error state when collection is added to unsupported context (#44879): When it's not possible to manually select a context, we will show an error state. We will ask merchant to change the collection using error state.



### How to test the changes in this Pull Request:



**Pre-requisite:** Install this plugin - 
[register-product-collection-tester.zip](https://github.com/user-attachments/files/16362323/register-product-collection-tester.zip)


**Case 1: `usesReference: ['product']` on Single Product template**

1. Go to "Single Product" template in Editor
2. Add "My Custom Collection - Product Context" block
3. Select the block
4. Verify that you can see preview label with message "Actual products will vary depending on the product being viewed." See screenshot below.

![image](https://github.com/user-attachments/assets/e296524f-ee08-4018-bed2-8a1b3adfe94b)

**Case 2: `usesReference: ['product']` at other places**

1. For other places i.e.:
  - Specific single product template
  - Other templates like Product Catalog, Product by Category etc.
  - Post or Page
2. Add "My Custom Collection - Product Context" block
3. Smoke test to verify that preview label isn't shown.
  - **Note:** These cases don't require a preview label. As part of #44877, we will show a prompt to select a product.

**Case 3: `usesReference: ['cart']`**

1. Go to "Page: Cart" template in Editor
2. Add "My Custom Collection - Cart Context" block
3. Select the block
4. Verify that you can see preview label with message "Actual products will vary depending on the page being viewed." See screenshot below.
5. Follow 1-4 steps for:
  - "Page: Checkout" template
  - Add "My Custom Collection - Cart Context" block inside "Cart" & "Mini cart" blocks
    - **Note:** I didn't mention Checkout block because i couldn't add "My Custom Collection - Cart Context" block inside Checkout block. But if you know how you can do it, please feel free to test with Checkout block as well.
6. For all other places, smoke test to verify that preview label isn't shown.

![image](https://github.com/user-attachments/assets/d4fa7a54-b44f-4204-af30-6d881cc6cf15)

**Case 4: `usesReference: ['order']`**

1. Go to "Order Confirmation" template in Editor
2. Add "My Custom Collection - Order Context" block
3. Select the block
4. Verify that you can see preview label with message "Actual products will vary depending on the page being viewed." 
5. For all other places, smoke test to verify that preview label isn't shown.

**Case 5: `usesReference: ['archive']`**

1. Go to any of the archive template in Editor i.e. Product by Category, Product by Tag, Products by Attribute etc.
2. Add "My Custom Collection - Archive Context" block
3. Select the block
4. Verify that you can see preview label with message "Actual products will vary depending on the page being viewed."
5. For all other places, smoke test to verify that preview label isn't shown.

**Case 6: `usesReference: ['product', 'order']` - Multiple values**

1. Go to "Single Product" template in Editor
2. Add "My Custom Collection - Multiple Contexts" block
3. Select the block
4. Verify that you can see preview label with message "Actual products will vary depending on the product being viewed."
5. Follow 1-4 steps for "Order Confirmation" template too.
6. For all other places, smoke test to verify that preview label isn't shown.

**Case 7: Advance testing (to be tested by developers only)**

Feel free to take a look at plugin code in `collections.js` file. 
- You can find all the collections of above cases in this file.
- Feel free to make some changes in collections and test how it works.



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 
Product Collection: Enable Context-Aware Previews by Adding `usesReference` to `registerProductCollection`
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
